arangod/Cluster/Maintenance.cpp

```diff
@@ -649,6 +663,77 @@ VPackBuilder getShardMap(VPackSlice const& collections) {
   return shardMap;
 }
 
+std::set<ShardID> extractShardsFromCollection(VPackSlice collection) {
+  auto shards = collection.get(SHARDS);
+  if (!shards.isObject()) {
+    return {};
+  }
+
+  std::set<ShardID> result;
+  for (auto shard : VPackObjectIterator(collection.get(SHARDS))) {
+    result.insert(ShardID::shardIdFromString(shard.key.stringView()).get());
+  }
+  return result;
+}
+
+std::tuple<DistributeShardsLikeMapping, ClonePrototypeMapping>
+getDistributeShardsLike(VPackSlice const& collections) {
+  DistributeShardsLikeMapping mapping;
+  ClonePrototypeMapping reverseMapping;
+
+  // maps a collection to the sorted set of shards
+  std::unordered_map<std::string_view, std::set<ShardID>> prototypeShards;
+
+  if (!collections.isObject()) {
+    return {};
+  }
+
+  for (auto collection : VPackObjectIterator(collections)) {
+    TRI_ASSERT(collection.value.isObject());
+
+    auto prototype = [&] {
+      if (auto s = collection.value.get("distributeShardsLike"); s.isString()) {
+        return s.stringView();
+      }
+      return std::string_view{};
+    }();
+
+    if (prototype.empty()) {
+      continue;
+    }
+    // check is we have collected information about this prototype already
+    auto it = prototypeShards.find(prototype);
+    if (it == prototypeShards.end()) {
+      auto protoCol = collections.get(prototype);
```

Comment:
nit: I had a hard time read `protoCol` because I was intuitively thinking of `protocol` (as in HTTP, HTTPS, etc.). I get that it means `protoCollection` here, but naming it `protoCol` is slightly ambiguous, so I suggest using `protoCollection` instead.